services/QuillLMS/spec/services/classlink_integration/enrollment_persister_spec.rb

```diff
@@ -0,0 +1,156 @@
+          ).run
+        end
+          .to not_change(ClassroomsTeacher, :count)
+          .and not_change(StudentsClassrooms, :count)
+      end
+    end
+
+    context 'association already exists' do
+      let(:classroom1) { create(:classroom) }
+      let!(:classlink_classroom1) do
+        create(
+          :classlink_classroom,
+          classroom: classroom1,
+          provider_instance: classlink_instance,
+          external_id: incoming_student_enrollment[:classroom_sourced_id]
+        )
+      end
+      let(:classroom2) { create(:classroom) }
+      let!(:classlink_classroom2) do
+        create(
+          :classlink_classroom,
+          classroom: classroom2,
+          provider_instance: classlink_instance
+        )
+      end
+
+      context 'role: teacher' do
+        let(:teacher1) { create(:teacher, role: 'coteacher') }
+
+        # This spec prevents a regression uncovered during integration testing
+        it 'should not raise a PG uniqueness constraint error' do
+          create(:classrooms_teacher, classroom: classroom1, user: teacher1, role: 'owner')
+
+          expect do
+            described_class.new(
+              incoming_enrollment: incoming_teacher_enrollment,
+              existing_provider_users: [],
+              existing_provider_classrooms: []
+            ).create_classroom_teacher(teacher1, classroom1)
+          end.to_not raise_error
+        end
+      end
+
+      context 'role: student' do
+        let(:student1) { create(:student) }
+        let!(:cl_account1) do
+          create(
+            :classlink_account,
+            user: student1,
+            classlink_instance:,
+            external_id: incoming_student_enrollment[:user_sourced_id]
+          )
+        end
+        let(:student2) { create(:student) }
+        let!(:cl_account2) do
+          create(
+            :classlink_account,
+            user: student2,
+            classlink_instance:
+          )
+        end
+
+        it 'should invoke StudentClassroomAssociator' do
+          existing_provider_users = ClasslinkAccount.where(
+            classlink_instance_id: classlink_instance.id
+          ).includes(:user)
+
+          existing_provider_classrooms = ClasslinkClassroom.where(
+            provider_instance_id: classlink_instance.id
+          ).includes(:classroom)
+          # TODO: refactor to use StudentClassroomsAssociator
+          # expect(StudentClassroomAssociator).to receive(:run).once
+          expect do
+            described_class.new(
+              incoming_enrollment: incoming_student_enrollment,
+              existing_provider_users:,
+              existing_provider_classrooms:
+            ).run
+          end
+            .to change(StudentsClassrooms, :count).by(1)
+            .and change(ClasslinkClassroomUser, :count).by 1
+        end
+      end
+    end
+
+    context 'association does not exist' do
+      context 'role: teacher' do
+        let(:existing_provider_users) do
```

Comment:
Good opportunity for one-liner:
`let(:existing_provider_users) { [create(:teacher)] }`